frontend/src/main/java/com/google/tripmeout/frontend/storage/PersistentPlaceVisitStorage.java

```diff
@@ -0,0 +1,140 @@
+    Transaction transaction = datastore.beginTransaction();
+    try {
+      datastore.get(placeKey);
+    } catch (EntityNotFoundException e) {
+      transaction.rollback();
+      throw new PlaceVisitNotFoundException("Place with id: " + placeVisitId + " doesn't exist");
+    }
+    datastore.delete(placeKey);
+    transaction.commit();
+  }
+
+  public Optional<PlaceVisitModel> getPlaceVisit(String tripId, String placeVisitId) {
+    Key placeKey = KeyFactory.stringToKey(placeVisitId);
+    Optional<PlaceVisitModel> optionalPlace = Optional.empty();
+    try {
+      Entity placeEntity = datastore.get(placeKey);
+      String mark = (String) placeEntity.getProperty(USER_MARK_PROPERTY_NAME);
+      PlaceVisitModel.UserMark userMark = PlaceVisitModel.UserMark.valueOf(mark);
+      optionalPlace = Optional.of(
+          PlaceVisitModel.builder()
+              .setId((String) placeEntity.getProperty(ID_PROPERTY_NAME))
+              .setPlaceName((String) placeEntity.getProperty(PLACE_NAME_PROPERTY_NAME))
+              .setPlacesApiPlaceId((String) placeEntity.getProperty(PLACE_API_ID_PROPERTY_NAME))
+              .setTripId((String) placeEntity.getProperty(TRIP_ID_PROPERTY_NAME))
+              .setUserMark(userMark)
+              .build());
+
+    } catch (EntityNotFoundException e) {
+    }
+    return optionalPlace;
+  }
+
+  public PlaceVisitModel updateUserMarkOrAddPlaceVisit(
+      PlaceVisitModel placeVisit, PlaceVisitModel.UserMark newStatus) {
+    Key placeKey = KeyFactory.stringToKey(placeVisit.id());
+    try {
+      Entity placeEntity = datastore.get(placeKey);
+      placeEntity.setProperty(USER_MARK_PROPERTY_NAME, newStatus.toString());
+      datastore.put(placeEntity);
+      placeVisit.toBuilder().setUserMark(newStatus);
+
+    } catch (EntityNotFoundException e) {
+    }
+    return placeVisit;
+  }
+
+  public List<PlaceVisitModel> getTripPlaceVisits(String tripId) {
+    Query query = new Query(tripId);
+    PreparedQuery results = datastore.prepare(query);
+    List<PlaceVisitModel> places = new ArrayList<>();
+    for (Entity placeEntity : results.asIterable()) {
+      String mark = (String) placeEntity.getProperty(USER_MARK_PROPERTY_NAME);
+      PlaceVisitModel.UserMark userMark = PlaceVisitModel.UserMark.valueOf(mark);
+      places.add(
+          PlaceVisitModel.builder()
+              .setId((String) placeEntity.getProperty(ID_PROPERTY_NAME))
+              .setPlaceName((String) placeEntity.getProperty(PLACE_NAME_PROPERTY_NAME))
+              .setPlacesApiPlaceId((String) placeEntity.getProperty(PLACE_API_ID_PROPERTY_NAME))
+              .setTripId((String) placeEntity.getProperty(TRIP_ID_PROPERTY_NAME))
+              .setUserMark(userMark)
+              .build());
+    }
+    return places;
+  }
+
+  public void removeTripPlaceVisits(String tripId) throws TripNotFoundException {
+    Query query = new Query(tripId);
+    PreparedQuery results = datastore.prepare(query);
```

Comment:
You'll need to do this in a transaction since you're querying and modifying the results separately. This also prevents the bamboozling case where this method returns successfully but there are actually remaining visits for the given trip (because they were inserted concurrently).